src/Algebra/Graph/Acyclic/Labelled/AdjacencyMap.hs

```diff
@@ -0,0 +1,352 @@
+-- License    : MIT (see the file LICENSE)
+-- Stability  : experimental
+--
+-- __Alga__ is a library for algebraic construction and manipulation of graphs
+-- in Haskell. See <https://github.com/snowleopard/alga-paper this paper> for the
+-- motivation behind the library, the underlying theory, and implementation details.
+--
+-- This module defines the 'AdjacencyMap' data type for edge-labelled acyclic graphs, as
+-- well as associated operations and algorithms.
+-----------------------------------------------------------------------------
+module Algebra.Graph.Acyclic.Labelled.AdjacencyMap (
+    -- * Data structure
+    AdjacencyMap, fromAcyclic,
+
+    -- * Basic graph construction primitives
+    empty, vertex, vertices,
+
+    -- * Relations on graphs
+    isSubgraphOf,
+
+    -- * Graph properties
+    isEmpty, hasVertex, hasEdge, edgeLabel, vertexCount, edgeCount, vertexList,
+    edgeList, vertexSet, edgeSet,
+
+    -- * Graph transformation
+    removeVertex, removeEdge, transpose, emap, induce, induceJust,
+
+    -- * Relational operations
+    transitiveClosure,
+
+    -- * Miscellaneous
+    -- consistent
+    ) where
+
+import Data.Set (Set)
+import Algebra.Graph.Label
+import Data.Function (on)
+
+import qualified Algebra.Graph.Labelled.AdjacencyMap as AM
+
+-- | Edge-labelled graphs, where the type variable @e@ stands for edge labels.
+-- For example, 'AdjacencyMap' @Bool@ @a@ is isomorphic to unlabelled acyclic graphs
+-- defined in the top-level module "Algebra.Graph.AdjacencyMap", where @False@
+-- and @True@ denote the lack of and the existence of an unlabelled edge,
+-- respectively.
+newtype AdjacencyMap e a = AAM {
+    -- | The /adjacency map/ of an edge-labelled graph: each vertex is
+    -- associated with a map from its direct successors to the corresponding
+    -- edge labels.
+    fromAcyclic :: AM.AdjacencyMap e a } deriving (Eq, Ord)
+
+-- TODO: Improve the Show instance.
+instance (Ord a, Show a, Ord e, Show e) => Show (AdjacencyMap e a) where
+  show x = "fromMaybe empty . toAcyclic $ " ++ show (fromAcyclic x)
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'AdjacencyMap'
+-- for more details.
+instance (Eq e, Dioid e, Ord a, Num a) => Num (AdjacencyMap e a) where
+  fromInteger = vertex . fromInteger
+  (AAM x) + (AAM y) = AAM $ induceEAM (uncurry (<)) (x + y)
+  (AAM x) * (AAM y) = AAM $ induceEAM (uncurry (<)) (x * y)
+  signum = const empty
+  abs = id
+  negate = id
+
+-- TODO: Add additional/change examples.
+
+-- TODO: Add the consistent function.
+
+-- TODO: Make 'skeleton :: Ord a => AdjacencyMap e a -> AAM.AdjacencyMap a'
+
+-- TODO: Add 'fromGraph'
+
+-- TODO: Add 'toAcyclic'
+
+-- | Construct the /empty graph/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'isEmpty'     empty == True
+-- 'hasVertex' x empty == False
+-- 'vertexCount' empty == 0
+-- 'edgeCount'   empty == 0
+-- @
+empty :: AdjacencyMap e a
+empty = AAM AM.empty
```

Comment:
This, as well as many other functions, can be implemented via `coerce` to ensure that they are zero-cost.

Reply:
Unfortunately, I am encountering an error when I use `coerce`,
![image](https://user-images.githubusercontent.com/17845944/61212238-4d0c7e80-a71f-11e9-8e7c-75c7fd01cdc0.png)
I tried writing my own `Coercible` instance but I get an "Illegal instance declaration" error.
I am still trying to figure out how to solve the error.

Reply:
See this comment for a similar issue and solution: https://github.com/snowleopard/alga/pull/220#issuecomment-510891564

Reply:
@snowleopard Thanks for the help. I've changed the required.